pkg/loop/prom_test.go

```diff
@@ -0,0 +1,56 @@
+package loop
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/stretchr/testify/require"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/logger"
+)
+
+func TestPromServer(t *testing.T) {
+
+	testReg := prometheus.NewRegistry()
+	testHandler := promhttp.HandlerFor(testReg, promhttp.HandlerOpts{})
+	testMetric := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "test_metric",
+	})
+	testReg.MustRegister(testMetric)
+	testMetric.Inc()
+
+	s := newPromServer(0, logger.Test(t), withHandler(testHandler))
+	// check that port is not resolved yet
+	require.Equal(t, -1, s.Port())
+	require.NoError(t, s.Start())
+
+	url := fmt.Sprintf("http://localhost:%d/metrics", s.Port())
+	resp, err := http.Get(url) //nolint
+	require.NoError(t, err)
+	require.NoError(t, err, "endpoint %s", url)
+	require.NotNil(t, resp.Body)
+	b, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	require.Contains(t, string(b), "test_metric")
+	defer resp.Body.Close()
+
+	require.NoError(t, s.Close())
+}
+
+// Port is the resolved port and is only known after Start().
+// returns -1 before it is resolved or if there was an error during resolution.
+func (p *promServer) Port() int {
```

Comment:
why is this defined in a test file? is that an oversight?

Reply:
It was only used in tests unless I missed something.

Reply:
I'll connect the other layers in drafts before merging this to be sure :shrug: